processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferArrayGrouper.java

```diff
@@ -0,0 +1,285 @@
+      aggregatorOffsets[i] = offset;
+      offset += aggregatorFactories[i].getMaxIntermediateSize();
+    }
+    recordSize = keySize + offset;
+  }
+
+  @Override
+  public void init()
+  {
+    if (!initialized) {
+      buffer = bufferSupplier.get();
+
+      reset();
+
+      initialized = true;
+    }
+  }
+
+  @Override
+  public boolean isInitialized()
+  {
+    return initialized;
+  }
+
+  @Override
+  public AggregateResult aggregate(KeyType key, int dimIndex)
+  {
+    Preconditions.checkArgument(dimIndex > -1, "Invalid dimIndex[%s]", dimIndex);
+
+    final ByteBuffer fromKey = keySerde.toByteBuffer(key);
+    if (fromKey == null) {
+      // This may just trigger a spill and get ignored, which is ok. If it bubbles up to the user, the message will
+      // be correct.
+      return Groupers.DICTIONARY_FULL;
+    }
+
+    if (fromKey.remaining() != keySerde.keySize()) {
+      throw new IAE(
+          "keySerde.toByteBuffer(key).remaining[%s] != keySerde.keySize[%s], buffer was the wrong size?!",
+          fromKey.remaining(),
+          keySerde.keySize()
+      );
+    }
+
+    final int recordOffset = dimIndex * recordSize;
+
+    if (recordOffset + recordSize > buffer.capacity()) {
+      // This error cannot be recoverd, and the query must fail
+      throw new ISE(
+          "A record of size [%d] cannot be written to the array buffer at offset[%d] "
+          + "because it exceeds the buffer capacity[%d]. Try increasing druid.processing.buffer.sizeBytes",
+          recordSize,
+          recordOffset,
+          buffer.capacity()
+      );
+    }
+
+    if (!isUsedKey(buffer, recordOffset)) {
+      this.buffer.position(recordOffset);
+      this.buffer.putInt(Groupers.getUsedFlag(dimIndex));
+      this.buffer.put(fromKey);
+    }
+
+    final int baseOffset = recordOffset + keySize;
+    for (int i = 0; i < aggregators.length; i++) {
+      aggregators[i].aggregate(buffer, baseOffset + aggregatorOffsets[i]);
+    }
+
+    return AggregateResult.ok();
+  }
+
+  @Override
+  public AggregateResult aggregate(KeyType key)
+  {
+    // BufferArrayGrouper is used only for dictionary-indexed single-value string dimensions.
+    // Here, the key contains the dictionary-encoded value of the grouping key
+    // and we use it as the index for the aggregation array.
+    final ByteBuffer fromKey = keySerde.toByteBuffer(key);
+    final int dimIndex = fromKey.getInt() + 1; // the first index is for missing value
+    fromKey.rewind();
+    return aggregate(key, dimIndex);
+  }
+
+  private static boolean isUsedKey(ByteBuffer buffer, int pos)
+  {
+    return (buffer.get(pos) & 0x80) == 0x80;
+  }
+
+  @Override
+  public void reset()
+  {
+    buffer.putInt(0, 0); // for missing value
+    for (int i = 1; i < cardinality + 1; i++) {
+      buffer.putInt(i * recordSize, 0);
+      final int baseOffset = i * recordSize + keySize;
+      for (int j = 0; j < aggregators.length; ++j) {
```

Comment:
Good point! Changed to lazy initialization. I run a benchmark and could see the lazy initialization improves performance quite significantly.